006_buildkite_importing_pipelines/README.md

```diff
@@ -0,0 +1,374 @@
+
+#### :arrow_right: Apply the schema
+
+We need to tell CUE to apply the schema to each pipeline.
+
+To do this we'll create a file at `internal/ci/buildkite/pipelines.cue` in our
+example.
+
+However, **if the pipeline imports that you performed earlier *already* created
+a file with that same path and name**, then simply select a different CUE
+filename that *doesn't* already exist. Place the file in the
+`internal/ci/buildkite/` directory.
+
+:floppy_disk: `internal/ci/buildkite/pipelines.cue`
+
+```
+package buildkite
+
+// each member of the pipelines struct must be a valid #Pipeline
+pipelines: [_]: #Pipeline
+```
+
+### Generate YAML from CUE
+
+#### :arrow_right: Create a CUE tool file
+
+Create a CUE "tool" file at `internal/ci/buildkite/ci_tool.cue` and adapt the
+element commented with `TODO`:
+
+:floppy_disk: `internal/ci/buildkite/ci_tool.cue`
+```CUE
+package buildkite
+
+import (
+	"path"
+	"encoding/yaml"
+	"tool/file"
+)
+
+_goos: string @tag(os,var=os)
+
+// Regenerate all pipeline files
+command: regenerate: {
+	pipeline_files: {
+		// TODO: update _toolFile to reflect the directory hierarchy containing this file.
+		let _toolFile = "internal/ci/buildkite/ci_tool.cue"
+		let _pipelineDir = path.FromSlash(".buildkite", path.Unix)
+		let _donotedit = "Code generated by \(_toolFile); DO NOT EDIT."
+
+		clean: {
+			glob: file.Glob & {
+				glob: path.Join([_pipelineDir, "*.yml"], _goos)
+				files: [...string]
+			}
+			for _, _filename in glob.files {
+				"Delete \(_filename)": file.RemoveAll & {path: _filename}
+			}
+		}
+
+		create: {
+			for _pipelineName, _pipeline in pipelines
+			let _filename = _pipelineName + ".yml" {
+				"Generate \(_filename)": file.Create & {
+					$after: [ for v in clean {v}]
+					filename: path.Join([_pipelineDir, _filename], _goos)
+					contents: "# \(_donotedit)\n\n\(yaml.Marshal(_pipeline))"
+				}
+			}
+		}
+	}
+}
+```
+
+Make the modification indicated by the `TODO` comment.
+
+This tool will export each CUE-based pipeline back into its required YAML file,
+on demand.
+
+#### :arrow_right: Test the CUE tool file
+
+With the modified `ci_tool.cue` file in place, check that the `regenerate`
+command is available **from a shell sitting at the repo root**. For example:
+
+:computer: `terminal`
+```sh
+cd $(git rev-parse --show-toplevel)               # make sure we're sitting at the repository root
+cue help cmd regenerate ./internal/ci/buildkite   # the "./" prefix is required
+```
+
+Your output **must** begin with the following:
+
+```text
+Regenerate all pipeline files
+Usage:
+  cue cmd regenerate [flags]
+[... output continues ...]
+```
+
+|   :exclamation: WARNING :exclamation:   |
```

Comment:
Just to confirm: if the command does not exist and we try to call it, is that not a sufficient test for its existence? Rather than also having this additional "run `help` and check output" step?

Reply:
So this warning block is the scar tissue from the *many* mistakes I've personally made whilst setting up and troubleshooting _tools!

There are a few error modes which trigger the `cue cmd foo` invocation to print the /generic/ `cue help cmd` long help text - with the context specific "why `foo` couldn't run / unify successfully / be found / etc etc etc" error preamble being stuck right at the top, invisible to the non-fastidious reader.

In my repeated experience, the only **definite** litmus test showing that things will work correctly is getting `cue help cmd foo`, to display the command-specific comment/text. That's why I use this formulation across multiple guides, and I call it out so explicitly in this warning block. I've found the _tool debugging process to be sufficiently *frustrating* in various situations that I wanted to set the reader up with the best possible chance of not simply continuing blindly, assuming that things will "just work" ...

Hence this warning and preceding step :-)

Reply:
Leaving this conversation unresolved, for a 👍 / 👎 .

Reply:
@myitcv Assuming a tacit 👍 for now, but leaving this unresolved.

Reply:
> There are a few error modes which trigger the `cue cmd foo` invocation to print the /generic/ `cue help cmd` long help text - with the context specific "why `foo` couldn't run / unify successfully / be found / etc etc etc" error preamble being stuck right at the top, invisible to the non-fastidious reader.

Makes sense. Thanks for the reminder. 

Please can you ensure we have an issue for this?

Reply:
Annoyingly, I don't have the triggers of this behaviour to hand! The next time I come across one I'll kick off an issue. (It'd feel slightly odd, starting an issue with "this sometimes happens, but I can't give you a hint as to /when/" ... :-))